ARCHITECTURE.md

```diff
@@ -0,0 +1,252 @@
+    F ->>  G: Respond with credential
+    G ->> -C: Respond with credential
+```
+
+The division into multiple services and APIs has two purposes:
+
+- least privilege: if there is a vulnerability in the UI process, it shouldn't
+  have access to interact with the credential service memory directly. Also,
+  hosting separate D-Bus services allows us to set different access control
+  policies on the bus, restricting communication on the Flow Control and UI
+  Control APIs to the `credentialsd` and `credentialsd-ui` services.
+
+- flexibility: UIs are very specific to their desktop environment, so it would
+  be impossible to satisfy the style and requirements for all the various Linux
+  desktop environments out there. This separation allows desktop environments to
+  create their own UIs targeted for their users.
+
+# Code Map
+
+## credentialsd
+
+A Rust binary project for the service hosting the Gateway and Flow Control APIs.
+Interacts with authenticators and clients/user agents.
+
+`credentialsd` does not start the UI directly; it sends a request to start the
+UI via D-Bus. It relies on D-Bus service activation or some other external
+method to start the process that hosts the UI Control API.
+
+Credential requests on the Gateway are handled one at a time; if a new request
+comes in, it is immediately rejected, and the client is expected to retry if
+necessary. Currently, clients cannot cancel their own requests; the user has to
+do that via the UI.
+
+### `credentialsd/src/credential_service/`
+
+`CredentialService` is the main component that interacts with authenticators. It
+also holds request context to return back to the Gateway for request completion
+or when the Flow Controller notifies it that the request is cancelled.
+
+Various authenticator transports are handled in sub-modules, for now USB and
+hybrid transports are supported. Each handler starts a `Stream` of events that
+represents requests from the authenticator for user interaction. If a response
+is required from the user, the event should contain a channel for the credential
+service to send the response after it receives user input.
+
+The credential service mostly just forwards events over to the UI service, minus
+any details that are not necessary for the UI to know (like the response
+channels mentioned above, which cannot be serialized over D-Bus anyway).
+
+Actual interaction I/O is performed in the [libwebauthn][libwebauthn] library.
+
+[libwebauthn]: https://github.com/linux-credentials/libwebauthn
+
+### `credentialsd/src/dbus/`
+
+D-Bus clients and services.
+
+The Gateway and Flow Controller services are defined here, as well as a client
+for the UI Controller.
+
+The `model` module contains some methods to convert from D-Bus types to internal
+credential service types. (These types don't need to be made known to the UI, so
+they do not live in `credentialsd-common`).
+
+### `credentialsd/src/webauthn.rs`
+
+Types and functions to deal with WebAuthn data.
```

Comment:
Could be a bit more precise. Something like "Types and functions needed to repackage requests from and responses to JSON-strings according to the WebAuthn spec (link: https://www.w3.org/TR/webauthn-3).
With one notable deviation from the spec: Due to D-Bus limitations, raw binary fields need to be base64 encoded strings. 
It is the responsibility of the application using this service to de/construct the field accordingly."